viewer/Viewer.cpp

```diff
@@ -750,25 +842,138 @@ namespace Enki
 					}
 				}
 				
-				
 				if (!found)
 					renderSimpleObject(*it);
 			}
 			
+			// draw object
 			glPushMatrix();
 			
 			glTranslated((*it)->pos.x, (*it)->pos.y, 0);
 			glRotated(rad2deg * (*it)->angle, 0, 0, 1);
 			
 			ViewerUserData* userData = polymorphic_downcast<ViewerUserData *>((*it)->userData);
+
 			userData->draw(*it);
 			displayObjectHook(*it);
 			
 			glPopMatrix();
 		}
-		
+	}
+
+	void ViewerWidget::picking(float left, float right, float bottom, float top, float zNear, float zFar)
+	{
+		pointedObject = 0;
+		QPoint cursorPosition = mapFromGlobal(QCursor::pos());
+
+		if (!rect().contains(cursorPosition,true)) // window don't contain cursor
+			return;
+
+		// prepare matricies for invertion
+		QMatrix4x4 projection;
+			projection.setToIdentity();
+			projection.frustum(left, right, bottom, top, zNear, zFar);
+		QMatrix4x4 modelview;
+			modelview.setToIdentity();
+			modelview.rotate(-90, 1, 0, 0);
+			modelview.rotate(rad2deg * -camera.pitch, 1, 0, 0);
+			modelview.rotate(90, 0, 0, 1);
+			modelview.rotate(rad2deg * -camera.yaw, 0, 0, 1);
+			modelview.translate(-camera.pos.x(), -camera.pos.y(), -camera.altitude);
+		QMatrix4x4 transformMatrix = (projection*modelview).inverted();
+
+		// cursor position in viewport coordinates
+		float fragmentX = (float)(cursorPosition.x() - width()/2)/(width()/2);
+		float fragmentY = (float)(height() - cursorPosition.y() - height()/2)/(height()/2);
+		float depth;
+		glReadPixels( cursorPosition.x(), height() - cursorPosition.y(), 1, 1, GL_DEPTH_COMPONENT, GL_FLOAT, &depth );
+
+		QVector4D input(fragmentX, fragmentY, 2*depth - 1, 1);
+		input = transformMatrix*input;
+
+		if (input.w() != 0.0) // valid pointed point
+		{
+			pointedPoint = QVector3D(input.x(),input.y(),input.z());
+			pointedPoint /= input.w();
+		}
+		else return;
+
+		// prepare to find which object is pointed
+		Point cursor2Dpoint(pointedPoint.x(),pointedPoint.y());
+		float cursorRadius = 0.05f;
+		for (World::ObjectsIterator it = world->objects.begin(); it != world->objects.end(); ++it)
+		{
+			const Vector distOCtoOC = (*it)->pos - cursor2Dpoint;		// distance between object bounding circle center and pointed point
+			const double addedRay = (*it)->getRadius() + cursorRadius;	// sum of bounded circle radius
+			if (distOCtoOC.norm2() <= (addedRay*addedRay)) 			// cursor point colide bounding circle
+			{
+				if(!(*it)->getHull().empty())				// check pointer circle and bject hull
```

Comment:
Enki's coding convention is to put a space between `if` and the opening parenthesis: `if (`.